src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -3638,6 +3644,29 @@
 
             matchedName = null;
             return (FirstIndex: -1, SecondIndex: -1);
+
+            void ensureCaseInsensitiveDictionary()
+            {
+                if (!_lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap.IsDefault)
+                {
+                    return;
+                }
+
+                if (_lazyForwardedTypesToAssemblyIndexMap.Count == 0)
+                {
+                    _lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap = ImmutableSegmentedDictionary<string, (int FirstIndex, int SecondIndex)>.Empty;
+                    return;
+                }
+
+                var caseInsensitiveMap = ImmutableSegmentedDictionary.CreateBuilder<string, (int FirstIndex, int SecondIndex)>(StringComparer.OrdinalIgnoreCase);
+
+                foreach (var (key, indexPair) in _lazyForwardedTypesToAssemblyIndexMap)
+                {
+                    _ = caseInsensitiveMap.TryAdd(key, indexPair);
+                }
+
+                _lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap = caseInsensitiveMap.ToImmutable();
```

Comment:
>_lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap = caseInsensitiveMap.ToImmutable(); [](http://example.com/codeflow?start=16&length=88)

Since ```ImmutableSegmentedDictionary``` is a struct, the assignment is no longer an atomic operation, and, therefore, is not thread-safe.

Reply:
Thanks for the catch.